Expecto/Expecto.fs

```diff
@@ -1613,6 +1626,83 @@ module Tests =
   /// The default configuration for Expecto.
   let defaultConfig = ExpectoConfig.defaultConfig
 
+  module Args =
+    open Microsoft.FSharp.Core
+
+    type Parser<'a> = (string[] * int * int) -> Result<'a,string> * int
+
+    let parseOptions (options:(string * string * Parser<_>) list) (strings:string[]) =
+      let rec updateUnknown unknown last length =
+        if length=0 then unknown
+        else updateUnknown (strings.[last]::unknown) (last-1) (length-1)
+      let rec collect help unknown args paramCount i =
+        if i>=0 then
+          let si = strings.[i]
+          if si = "--help" then
+            collect true (updateUnknown unknown (i+paramCount) paramCount) args 0 (i-1)
+          else
+            match List.tryFind (fst3 >> (=)si) options with
+            | Some(option,_,parser) ->
+              let arg, unknownCount = parser (strings,i+1,paramCount)
+              collect help
+                (updateUnknown unknown (i+paramCount) unknownCount)
+                (Result.mapError (fun i -> option + " " + i) arg::args) 0 (i-1)
+            | None -> collect help unknown args (paramCount+1) (i-1)
```

Comment:
A future PR could do fuzzy matching of input here

Reply:
Yes a bit like powershell

Reply:
Powershell lets you shorten switches as long as they are unambiguous. This actually may not be a good idea for back compact reasons when new switches are added.

A possibly more useful one would be to suggest the correct switch based on a distance measure. Worth considering in the future. If there are enough good ideas I may pull Args out as a single file Args utility. Personally keen on where this could go as I use these quite a lot.

Reply:
Got it.  I think the PowerShell folks got it wrong and they are starting to realize it.  With PowerShell 6, a group in Microsoft was looking into Query Auto Completion-like features to predict what you want to type for a command.  Sort of like next-gen IntelliSense for the command line. https://microsoft.github.io/prose/ Microsoft Program Synthesis using Examples SDK

Can you share how you use it a lot?

Simplest way to share would be to create a PowerShell Profile with a History logger extension so that you can record what you do in each session.  Then you can really data mine your use cases.